airgen/src/lib.rs

```diff
@@ -295,11 +309,128 @@ impl<'a> ASMPILConverter<'a> {
                         id: d.operation.id.id.clone(),
                         params: d.operation.params.clone(),
                     },
-                    selector_idx,
+                    // this will be set later, after compatible links are merged
+                    selector_idx: None,
                 })
                 .unwrap()
                 .clone(),
             is_permutation,
         }
     }
+
+    /// Process each link and then combine compatible links.
+    /// Links can be merged iff:
+    /// - they originate from the same machine instance
+    /// - they target the same instance.operation
+    /// - they are of the same kind (permutation/lookup)
+    /// - their flags are mutually exclusive
+    /// Right now we only consider links from different instructions,
+    /// as a single instruction can be active at a time
+    fn process_and_merge_links(&self, defs: &[LinkDefinition]) -> Vec<Link> {
+        /// Helper struct to group links that can potentially be merged.
+        /// Besides these being equal, the links must be mutually exclusive (e.g., come from different instructions)
+        #[derive(Clone, Ord, PartialOrd, Eq, PartialEq, Debug)]
+        struct LinkInfo {
+            from: Location,
+            to: Location,
+            operation: Operation,
+            is_permutation: bool,
+        }
+
+        // resulting links
+        let mut links = vec![];
+
+        // idea here is:
+        // - group by LinkInfo
+        // - inside each group, separate links into sets of with mutually exclusive flags (that is, from different instructions)
+        let mut grouped_links: BTreeMap<LinkInfo, Vec<BTreeMap<Expression, Link>>> =
+            Default::default();
+        for link_def in defs.iter() {
+            let link = self.handle_link_def(link_def.clone());
+
+            // group compatible links
+            if link.from.instr_flag.is_some() {
+                let info = LinkInfo {
+                    from: self.location.clone(),
+                    to: link.to.machine.location.clone(),
+                    operation: link.to.operation.clone(),
+                    is_permutation: link.is_permutation,
+                };
+
+                // TODO: temporary solution to a witgen limitation: avoid merging links that use next references.
+                // Remove this when witgen supports it (and update related tests).
+                use powdr_ast::parsed::{
+                    visitor::ExpressionVisitable, UnaryOperation, UnaryOperator::Next,
+                };
+                use Expression::UnaryOperation as EUnaryOperation;
+                if link.from.params.inputs_and_outputs().any(|p| {
+                    p.expr_any(|e| matches!(e, EUnaryOperation(_, UnaryOperation { op: Next, .. })))
```

Comment:
```suggestion
                    p.expr_any(|e| e.contains_next_ref())
```

Reply:
But please check if expr_any already traverses the tree (beacuse contains_next_ref also does)

Reply:
Actually looked for such a function and didn't find it! Didn't find it because it only exists for `AlgebraicExpression` and this is a `parsed::Expression`... I'll create it